networkx/algorithms/bipartite/link_analysis.py

```diff
@@ -243,17 +243,22 @@ def birank(
     p0 = np.array([top_personalization.get(n, 0) for n in top], dtype=float)
     u0 = np.array([bottom_personalization.get(n, 0) for n in bottom], dtype=float)
 
-    # Construct degree normalized biadjacency matrix `S`
+    # Construct degree normalized biadjacency matrix `S` and its transpose
     W = nx.bipartite.biadjacency_matrix(G, bottom, top, weight=weight, dtype=float)
     D_p = sp.sparse.dia_array(
-        ([1.0 / np.sqrt(W.sum(axis=0))], [0]), shape=(top_count, top_count), dtype=float
+        ([1.0 / np.sqrt(W.sum(axis=0))], [0]),
+        shape=(top_count, top_count),
+        dtype=float,
     )
     D_u = sp.sparse.dia_array(
         ([1.0 / np.sqrt(W.sum(axis=1))], [0]),
         shape=(bottom_count, bottom_count),
         dtype=float,
     )
+    D_u.tocsr()
+    D_p.tocsr()
     S = D_u @ W @ D_p
```

Comment:
Actually `D_U.tocsr()` returns a CSR, the old DIA is still untouched. So since D_u and D_p are not used again, you could use `S = D_U.tocsr() @ W @ D_p.tocsr()`. Or you could convert like `D_u = D_u.tocsr()`.  But that actually isn't any better performance-wise than what scipy does under the hood. Their philosophy is that the user shouldn't need to worry about what format it is -- all the conversions are linear in size of array so we can leave it to them to pick which is the best format for e.g. matmul.

So lots of ways to do this -- but the current version converts and then doesn't assign it to anything. :)

Reply:
Yep! Switched to `S = D_u.tocsr() @ W @ D_p.tocsr()`.